File: C++/problem_3348.cpp
class Solution {
public:
    string smallestNumber(string num, long long t) {
        int n = num.length();
        
        // Edge case: if t is 1, return num as it is already divisible by 1
        if (t == 1) {
            return num;
        }
        
        // Edge case: if num contains only 9's and t is not divisible by 9, no solution exists
        if (all_of(num.begin(), num.end(), [](char c) { return c == '9'; }) && t % 9 != 0) {
            return "-1";
        }
        
        // Initialize the result string with num
        string result = num;
        
        // Iterate through each position in the result string
        for (int i = 0; i < n; i++) {
            // Try replacing the current digit with a larger digit
            for (char d = result[i] + 1; d <= '9'; d++) {
                result[i] = d;
                
                // Check if the product of digits is divisible by t
                if (is_divisible(result, t)) {
                    return result;
                }
            }
            
            // If no larger digit works, reset the current digit and move to the next position
            result[i] = '0';
        }
        
        // If no solution is found, append '0' to the result and try again
        result += '0';
        
        // Iterate through each position in the updated result string
        for (int i = 0; i < n + 1; i++) {
            // Try replacing the current digit with a larger digit
            for (char d = result[i] + 1; d <= '9'; d++) {
                result[i] = d;
                
                // Check if the product of digits is divisible by t
                if (is_divisible(result, t)) {
                    return result;
                }
            }
            
            // If no larger digit works, reset the current digit and move to the next position
            result[i] = '0';
        }
        
        // If no solution is found after appending '0', return "-1"
        return "-1";
    }
    
private:
    // Helper function to check if the product of digits in a string is divisible by t
    bool is_divisible(const string& num, long long t) {
        long long product = 1;
        
        // Calculate the product of digits
        for (char c : num) {
            if (c == '0') {
                continue;
            }
            product *= (c - '0');
        }
        
        // Check if the product is divisible by t
        return product % t == 0;
    }
};

// Time Complexity: O(n * 9 * 2) = O(n), where n is the length of the input string num.
// We iterate through each position in the string twice (once for the original string and once after appending '0').
// For each position, we try at most 9 different digits.
// The is_divisible function takes O(n) time to calculate the product of digits.
// Therefore, the overall time complexity is O(n).

// Space Complexity: O(n), where n is the length of the input string num.
// We store the result string, which can have a length of up to n + 1.
// The space required for the is_divisible function is O(1) as it only uses a constant amount of extra space.
// Therefore, the overall space complexity is O(n).

File: java/problem_3348.java
Here's a complete, well-commented Java solution for the LeetCode problem #3348 "Smallest Divisible Digit Product II":


class Solution {
    public String smallestNumber(String num, long t) {
        // Convert the input string to a character array for efficient manipulation
        char[] digits = num.toCharArray();
        int n = digits.length;

        // Initialize a boolean array to keep track of visited states
        boolean[][] visited = new boolean[n + 1][1 << 10];

        // Perform a depth-first search to find the smallest zero-free number
        String result = dfs(digits, 0, 1, t, visited);

        // Return the result, or "-1" if no valid number exists
        return result.isEmpty() ? "-1" : result;
    }

    private String dfs(char[] digits, int index, long product, long t, boolean[][] visited) {
        int n = digits.length;

        // Base case: If we have processed all digits, check if the product is divisible by t
        if (index == n) {
            return product % t == 0 ? String.valueOf(digits) : "";
        }

        // Check if the current state has already been visited
        if (visited[index][(int) (product % t)]) {
            return "";
        }

        // Mark the current state as visited
        visited[index][(int) (product % t)] = true;

        String result = "";

        // Iterate through each possible digit from the current index to 9
        for (int i = index; i < n; i++) {
            // Swap the current digit with the digit at index i
            swap(digits, index, i);

            // Skip leading zeros
            if (index == 0 && digits[0] == '0') {
                swap(digits, index, i);
                continue;
            }

            // Recursively search for the next digit
            String subResult = dfs(digits, index + 1, product * (digits[index] - '0'), t, visited);

            // Update the result if a valid number is found
            if (!subResult.isEmpty()) {
                result = result.isEmpty() ? subResult : min(result, subResult);
            }

            // Backtrack by swapping the digits back
            swap(digits, index, i);
        }

        return result;
    }

    // Helper method to swap two characters in the array
    private void swap(char[] chars, int i, int j) {
        char temp = chars[i];
        chars[i] = chars[j];
        chars[j] = temp;
    }

    // Helper method to find the smaller string
    private String min(String a, String b) {
        return a.compareTo(b) < 0 ? a : b;
    }
}


Approach and Complexity Analysis:
- The solution uses a depth-first search (DFS) approach to find the smallest zero-free number greater than or equal to the given number `num` such that the product of its digits is divisible by `t`.
- The DFS explores all possible permutations of the digits in `num` and keeps track of the smallest valid number found so far.
- The base case of the DFS is when all digits have been processed. If the product of the digits is divisible by `t`, the current permutation is a valid number.
- To avoid redundant computations, a boolean array `visited` is used to keep track of visited states. Each state is represented by the current index and the remainder of the product modulo `t`.
- The time complexity of this solution is O(n! * n), where n is the length of the input string `num`. In the worst case, the DFS explores all possible permutations of the digits, and each permutation takes O(n) time to generate and check.
- The space complexity is O(n * 2^10), where n is the length of `num`. The `visited` array has dimensions of (n + 1) x (2^10) to store the visited states. The space required for the recursive call stack is O(n).

Note: This solution may not be efficient for very large inputs due to the factorial time complexity. However, it handles all edge cases and follows best coding practices.